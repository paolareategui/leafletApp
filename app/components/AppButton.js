import React from "react";

import { StyleSheet, Text, TouchableOpacity, View } from "react-native";

import AppColors from "../config/AppColors";

function AppButton({ color = "secondaryColor", title, onPress }) {
  return (
    <TouchableOpacity onPress={onPress} style={styles.container}>
      <View style={[styles.button, { backgroundColor: AppColors[color] }]}>
        <Text style={styles.buttonTitle}>{title}</Text>
      </View>
    </TouchableOpacity>
  );
}

const styles = StyleSheet.create({
  container: {
    alignItems: "center",
  },
  button: {
    padding: 15,
    width: "70%",
    borderRadius: 8,
  },
  buttonTitle: {
    fontSize: 17,
    fontWeight: "600",
    color: AppColors.white,
    textTransform: "uppercase",
    textAlign: "center",
  },
});

export default AppButton;
