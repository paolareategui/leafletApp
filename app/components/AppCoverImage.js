import React from "react";

import { Image, Text, StyleSheet } from "react-native";

import AppColors from "../config/AppColors";

function AppCoverImage({ source }) {
  return <Image style={styles.container} source={source} />;
}

const styles = StyleSheet.create({
  container: {
    resizeMode: "cover",
    width: 390,
    height: 390,
  },
});

export default AppCoverImage;
