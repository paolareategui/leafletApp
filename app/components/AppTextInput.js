import React from "react";

import { StyleSheet, TextInput, View } from "react-native";

import AppColors from "../config/AppColors";

function AppTextInput({ color = AppColors.black, ...otherProps }) {
  return (
    <View style={styles.container}>
      <TextInput
        style={styles.textInput}
        placeholderTextColor={color}
        {...otherProps}
      />
    </View>
  );
}

const styles = StyleSheet.create({
  container: {
    marginTop: 30,
  },

  textInput: {
    backgroundColor: AppColors.white,
    width: "100%",
    alignSelf: "center",
    padding: 15,
    borderColor: AppColors.secondaryColor,
    borderWidth: 0.5,
    borderRadius: 4,
  },
});

export default AppTextInput;
