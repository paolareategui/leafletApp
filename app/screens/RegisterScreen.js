import React, { useState } from "react";

import { Formik } from "formik";
import * as Yup from "yup";

import AppErrorText from "../components/AppErrorText";
import users from "../data/users.json";

import {
  KeyboardAvoidingView,
  StyleSheet,
  Text,
  TouchableOpacity,
  View,
} from "react-native";

import AppButton from "../components/AppButton";
import AppCoverImage from "../components/AppCoverImage";
import AppText from "../components/AppText";
import AppTextInput from "../components/AppTextInput";

//Yup schema for validation
const schema = Yup.object().shape({
  username: Yup.string().required().min(2).label("username"),
  email: Yup.string().required().email().label("Email"),
  password: Yup.string().required().min(4).max(8).label("Password"),
});

function RegisterScreen({ navigation }) {
  //Check if userid already exists
  const checkUser = ({ email }) => {
    return users.filter((user) => user.email === email).length > 0;
  };
  //Add new user instance
  const addUser = ({ email, password, username }) => {
    //Check if user already exists
    const userExists = checkUser({ email });
    console.log("Does user exist?", userExists);

    //Do accordinly
    if (userExists) {
      alert("You already have an account");
    } else {
      const newUser = {
        userid: email,
        username: username,
        email: email,
        password: password,
      };
      //Add new user to user array
      users.push(newUser);
      //Message and redirection for new user
      alert("Thanks for registering!");
      navigation.navigate("Login");
    }
  };

  return (
    //Everything inside KeyboardAvoidingView will move up when the keyboard appears
    <KeyboardAvoidingView keyboardVerticalOffset={100} behavior="position">
      {/* Cover image */}
      <View style={styles.coverImageView}>
        <AppCoverImage source={require("../assets/welcome.png")} />
      </View>

      {/* Sign up form starts */}
      <Formik
        initialValues={{ username: "", email: "", password: "" }}
        onSubmit={(values, { resetForm }) => {
          addUser(values);
          resetForm({});
          console.log(values);
        }}
        validationSchema={schema}
      >
        {/* Variables to be passed to form */}
        {({
          errors,
          handleChange,
          handleSubmit,
          setFieldTouched,
          touched,
          values,
        }) => (
          <>
            <View style={styles.textInputContainer}>
              {/* username */}
              <AppTextInput
                autoCapitalize="none"
                autoCorrect={false}
                clearButtonMode="while-editing"
                keyboardType="default"
                onBlur={() => setFieldTouched("username")}
                onChangeText={handleChange("username")}
                placeholder="Your name"
                textContentType="username"
                value={values.username || ""}
              />

              {/* Error message for username */}
              {touched.username && (
                <AppErrorText>{errors.username}</AppErrorText>
              )}

              {/* Email */}
              <AppTextInput
                autoCapitalize="none"
                autoCorrect={false}
                clearButtonMode="while-editing"
                keyboardType="email-address"
                onBlur={() => setFieldTouched("email")}
                onChangeText={handleChange("email")}
                placeholder="Email"
                textContentType="emailAddress"
                value={values.email || ""}
              />

              {/* Error message for email */}
              {touched.email && <AppErrorText>{errors.email}</AppErrorText>}

              {/* Password */}
              <AppTextInput
                autoCapitalize="none"
                autoCorrect={false}
                onBlur={() => setFieldTouched("password")}
                clearButtonMode="while-editing"
                keyboardType="name-phone-pad"
                onChangeText={handleChange("password")}
                placeholder="Password"
                secureTextEntry
                textContentType="password"
                value={values.password || ""}
              />

              {/* Error message for password */}
              {touched.password && (
                <AppErrorText>{errors.password}</AppErrorText>
              )}
            </View>

            <View style={styles.buttonContainer}>
              <AppButton title="SIGN UP" onPress={handleSubmit} />
            </View>
          </>
        )}
      </Formik>
      {/* Sign up Form ends */}

      {/* Link to signup form for returning users */}
      <View style={styles.loginLink}>
        <TouchableOpacity onPress={() => navigation.navigate("Login")}>
          <AppText style={{ fontSize: 13 }}>
            Already have an account?{" "}
            <Text style={{ fontWeight: "600" }}>Log in</Text>
          </AppText>
        </TouchableOpacity>
      </View>
    </KeyboardAvoidingView>
  );
}

const styles = StyleSheet.create({
  buttonContainer: {
    marginTop: 30,
  },
  coverImageView: {
    marginTop: "-20%",
  },
  loginLink: { alignItems: "center", marginTop: 40, marginBottom: 10 },
  textInputContainer: { alignSelf: "center", width: "70%" },
});

export default RegisterScreen;
