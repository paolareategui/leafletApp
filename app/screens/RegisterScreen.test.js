import React from "react";
import renderer from "react-test-renderer";
import AppButton from "../components/AppButton";
import AppCoverImage from "../components/AppCoverImage";
import AppTextInput from "../components/AppTextInput";
import RegisterScreen from "./RegisterScreen";

describe("Elements in RegisterScreen render properly", () => {
  test("The correct Cover image is displayed", () => {
    const navigation = jest.fn();
    const getTheProps = RegisterScreen({ navigation });
    //get the source url from props
    const propImgsrc =
      getTheProps.props.children[0].props.children.props.source;
    //the correct source url
    const imgSource = require("../assets/welcome.png");
    //compare urls
    expect(propImgsrc).toBe(imgSource);
  });

  test("The cover image has the right dimensions", () => {
    const tree = renderer.create(<AppCoverImage />).toJSON();
    expect(tree.props.style.resizeMode).toBe("cover");
    expect(tree.props.style.width).toBe(390);
    expect(tree.props.style.height).toBe(390);
  });

  test("<AppTextInput /> renders correctly", () => {
    const tree = renderer.create(<AppTextInput />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test("Placeholder texts are correct", () => {
    const tree = renderer.create(<RegisterScreen />).toJSON();
    const namePlaceholder =
      tree.children[0].children[1].children[0].children[0].props.placeholder;
    const emailPlaceholder =
      tree.children[0].children[1].children[1].children[0].props.placeholder;
    const passPlaceholder =
      tree.children[0].children[1].children[2].children[0].props.placeholder;
    //check name placeholder
    expect(namePlaceholder).toBe("Your name");
    //check email placeholder
    expect(emailPlaceholder).toBe("Email");
    //check password placeholder
    expect(passPlaceholder).toBe("Password");
  });

  test("<AppButton /> renders correctly", () => {
    const tree = renderer.create(<AppButton />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test("the label for the submit button is SIGN UP", () => {
    const tree = renderer.create(<RegisterScreen />).toJSON();
    const theProps =
      tree.children[0].children[2].children[0].children[0].children[0]
        .children[0];
    expect(theProps).toBe("SIGN UP");
  });
});
