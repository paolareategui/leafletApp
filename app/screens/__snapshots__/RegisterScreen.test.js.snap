// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Elements in RegisterScreen render properly <AppButton /> renders correctly 1`] = `
<View
  accessible={true}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "alignItems": "center",
      "opacity": 1,
    }
  }
>
  <View
    style={
      Array [
        Object {
          "borderRadius": 8,
          "padding": 15,
          "width": "70%",
        },
        Object {
          "backgroundColor": "#16182B",
        },
      ]
    }
  >
    <Text
      style={
        Object {
          "color": "#fff",
          "fontSize": 17,
          "fontWeight": "600",
          "textAlign": "center",
          "textTransform": "uppercase",
        }
      }
    />
  </View>
</View>
`;

exports[`Elements in RegisterScreen render properly <AppTextInput /> renders correctly 1`] = `
<View
  style={
    Object {
      "marginTop": 30,
    }
  }
>
  <TextInput
    allowFontScaling={true}
    placeholderTextColor="#333"
    rejectResponderTermination={true}
    style={
      Object {
        "alignSelf": "center",
        "backgroundColor": "#fff",
        "borderColor": "#16182B",
        "borderRadius": 4,
        "borderWidth": 0.5,
        "padding": 15,
        "width": "100%",
      }
    }
    underlineColorAndroid="transparent"
  />
</View>
`;
